### PR TITLE
Linked subsystems.

### PR DESCRIPTION
**Pull request by VytalyGorbatov**
Dev branch is a branch where Services, Sockets, Logger and ThreadPool is merged and working fine. Also MVS19 solution is configured quite well, so work with this branch is much more convenient.

### CODE REVIEW

File: Sockets/BackEndMonitoringSockets/PropertySheets/IncludeSheetRelease64.props [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <ImportGroup Label="PropertySheets" />
+  <PropertyGroup Label="UserMacros" />
+  <PropertyGroup />
+  <ItemDefinitionGroup>
+    <Link>
+      <AdditionalLibraryDirectories>$(SolutionDir)packages\CLogger\$(Configuration)\x64\; %(AdditionalLibraryDirectories)</AdditionalLibraryDirectories>
+      <AdditionalDependencies>CLogger.lib;kernel32.lib;user32.lib;gdi32.lib;winspool.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Comment (AntonVoytov) [inline Sockets/BackEndMonitoringSockets/PropertySheets/IncludeSheetRelease64.props:9]: Looks like there is too many lib dependencies, some of them are not might needed. Please check other places.

**Review by AntonVoytov — COMMENTED** [review]
Please check comment with lib dependencies.

File: Sockets/BackEndMonitoringSockets/PropertySheets/IncludeSheetRelease64.props (hunk 1 above)
Comment (VytalyGorbatov) [reply Sockets/BackEndMonitoringSockets/PropertySheets/IncludeSheetRelease64.props]: Yes, sorry, this was miscopied. Edited it in [commit](https://github.com/ITA-Dnipro/BackEndMonitoring/pull/28/commits/74fd40f3cec3740749b2632ea0a8e42f7e3dc2df).

File: Server/BackEndMonitoringServer/src/BackEndMonitoringServer.cpp [hunk 2]
```diff
@@ -1,27 +1,23 @@
...
+    {
+        Utils::DisplayMessage("Invalid parameters");
+        return EXIT_FAILURE;
+    }
 
-	acceptor.StartServer();
-	system("pause");
-	return 0;
-}
+    const int return_code = success ? 0 : 1;
```
Comment (Chupakabra0) [inline Server/BackEndMonitoringServer/src/BackEndMonitoringServer.cpp:20]: If `!success` is `true`, then you step into if statement and `EXIT_FALIURE` will be returned. So you'll never get `return_code` = 1 in this situation. You can safely return `EXIT_SUCCESS`.

File: Server/BackEndMonitoringServer/src/CServiceHandler.cpp [hunk 3]
```diff
@@ -0,0 +1,287 @@
+#include "stdafx.h"
+
+#include "CServiceHandler.h"
+#include "Utils.h"
+
+class ServiceHandle
```
Comment (Chupakabra0) [inline Server/BackEndMonitoringServer/src/CServiceHandler.cpp:6]: Don't forget to decide what you have to do with move and copy ctors in future. For example, you can safely delete copy and leave default move ctors in this class.

File: Server/BackEndMonitoringServer/include/CService.h [hunk 4]
```diff
@@ -0,0 +1,64 @@
+#pragma once
+
+#include "CEvent.h"
+#include "Sockets/BackEndMonitoringSockets/include/CAcceptorWrapper.h"
+
+struct ServiceParameters
+{
+    const CString& name         = "BackendMonitoringService";
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/include/CService.h:8]: we don't need such formatted lines according to code style

File: Server/BackEndMonitoringServer/include/CService.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    static CService* m_service;
+    CString m_name;
+    CString m_display_name;
+    DWORD m_start_type;
+    DWORD m_error_control_type;
+    SERVICE_STATUS m_status;
+    SERVICE_STATUS_HANDLE m_status_handle;
+    std::thread m_main_thread;
+    CEvent m_stop_event;
+    std::shared_ptr<CThreadPool> m_thread_pool;
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/include/CService.h:62]: it's pointer. Must be m_p_thread_pool

File: Server/BackEndMonitoringServer/include/Utils.h [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#pragma once
+
+namespace Utils
+{
+
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/include/Utils.h:5]: don't need empty line

File: Server/BackEndMonitoringServer/src/CCommandLineHandler.cpp [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                success = service_handler->Uninstall();
+                break;
+            }
+
+            if (strcmp(argv[1], "help") == 0)
+            {
+                // Show help
+                break;
+            }
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/src/CCommandLineHandler.cpp:47]: You could put if -> else if -> else - then if made a mistake, help command will be shown

File: Server/BackEndMonitoringServer/src/CService.cpp [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+    std::fstream stream("Log.txt", std::ios_base::out);
+    CLogBuilder builder("Logger", ELogLevel::DEBUG_LEVEL);
+    builder.AddThreadUnsafeStream(stream).SetLogConfig(ELogConfig::CALL_TIME,
+        ELogConfig::FILE_NAME, ELogConfig::FUNCTION_NAME,
+        ELogConfig::LINE_NUMBER, ELogConfig::MESSAGE, ELogConfig::PARAMS);
+    auto logger = builder.BuildSharedLog();
+
+    size_t num_threads = 20;
+    int port = 1111;
+    std::string ip_address = "127.0.0.1";
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/src/CService.cpp:22]: maybe better to write comment that here will be XMLParser used?

File: Server/BackEndMonitoringServer/src/CService.cpp [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+const DWORD CService::GetStartType() const
+{ 
+    return m_start_type;
+}
+
+const DWORD CService::GetErrorControlType() const
+{
+    return m_error_control_type;
+}
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/src/CService.cpp:120]: formate to { return ...; }

File: Server/BackEndMonitoringServer/src/CServiceHandler.cpp [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+{
+public:
+    explicit ServiceHandle(SC_HANDLE handle)
+      : m_handle(handle) {}
+
+    ~ServiceHandle() 
+    {
+        if (m_handle != nullptr) 
+        {
+            if(::CloseServiceHandle(m_handle) == 0)
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/src/CServiceHandler.cpp:16]: swap places: nullptr != ....

File: Server/BackEndMonitoringServer/src/CServiceHandler.cpp [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        {
+            Utils::DisplayError("Failed to get module file name");
+            success = false;
+            break;
+        }
+
+        escaped_path.ReleaseBuffer();
+        escaped_path.Remove('\"');
+
+        escaped_path = '\"' + escaped_path + '\"';
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/src/CServiceHandler.cpp:58]: maybe better write fucntion FormatePath(string)?

File: Server/BackEndMonitoringServer/src/CServiceHandler.cpp [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+    
+    success ? Utils::DisplayMessage("Service restarted")
+            : Utils::DisplayMessage("Failed to restart the service");
+
+    return success;
+}
+
+bool ServiceHandler::Run() const
+{
+    return m_service->Run();
```
Comment (Crowy-o) [inline Server/BackEndMonitoringServer/src/CServiceHandler.cpp:286]: I couldn't find header, but, as I see, m_service - pointer. Then rename it into m_p_service

File: Sockets/BackEndMonitoringSockets/include/CAcceptorWrapper.h [hunk 13]
```diff
@@ -10,18 +10,18 @@ class CAcceptorWrapper
...
 		const std::string& address);
 	std::unique_ptr<CServiceConnectionHandler> InitServiceHandler(int handle);
 	std::unique_ptr<CSocketWrapper> InitSocketWrapper(int handle);
 
 	CEvent& m_event;
-	std::unique_ptr<CThreadPool> m_pool;
+	std::shared_ptr<CThreadPool> m_pool;
 	std::unique_ptr<CAcceptor> m_server_acceptor;
 	std::unique_ptr<CServiceConnectionHandler> m_service_handler;
 	std::unique_ptr<CSocketWrapper> m_stream;
```
Comment (Crowy-o) [inline Sockets/BackEndMonitoringSockets/include/CAcceptorWrapper.h:27]: if pointer - add prefix p_ (example: m_p_pool)

File: Sockets/BackEndMonitoringSockets/src/CAcceptor.cpp [hunk 14]
```diff
@@ -18,6 +18,11 @@ int CAcceptor::GetHandle() const
 	return m_peer_acceptor->GetHandle();
 }
 
+bool CAcceptor::CloseSocket()
+{
+	return m_peer_acceptor->CloseSocket();
+}
```
Comment (Crowy-o) [inline Sockets/BackEndMonitoringSockets/src/CAcceptor.cpp:24]: formate as 1 line { ..... }

File: Sockets/BackEndMonitoringSockets/src/CAcceptorWrapper.cpp [hunk 15]
```diff
@@ -2,38 +2,40 @@
...
 CAcceptorWrapper::CAcceptorWrapper(int port, const std::string& ip_address, 
-	size_t num_threads, CEvent& event, std::shared_ptr<CLogger> logger)
-	: m_event(event), m_logger(logger)
+	CEvent& event, std::shared_ptr<CThreadPool> pool, 
+	std::shared_ptr<CLogger> logger)
+	: m_event(event), m_pool(pool), m_logger(logger)
 {
 	m_server_acceptor = InitAcceptor(1111, "127.0.0.1");
 	m_stream = InitSocketWrapper(m_server_acceptor->GetHandle());
 	m_service_handler = InitServiceHandler(m_server_acceptor->GetHandle());
```
Comment (Crowy-o) [inline Sockets/BackEndMonitoringSockets/src/CAcceptorWrapper.cpp:11]: The constructor shouldn't implement hard init. Better to add another method for such initialization and added variable that check is_initialized object

File: Server/BackEndMonitoringServer/src/CCommandLineHandler.cpp (hunk 7 above)
Comment (AntonVoytov) [reply Server/BackEndMonitoringServer/src/CCommandLineHandler.cpp:47]: The better option is to change 'if else' on 'switch'

File: Server/BackEndMonitoringServer/src/BackEndMonitoringServer.cpp (hunk 2 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/src/BackEndMonitoringServer.cpp]: Thnx, fixed.

File: Server/BackEndMonitoringServer/src/CServiceHandler.cpp (hunk 3 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/src/CServiceHandler.cpp:6]: Fixed too.

File: Server/BackEndMonitoringServer/include/CService.h (hunk 4 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/include/CService.h]: Fixed, thnx.

File: Server/BackEndMonitoringServer/include/CService.h (hunk 5 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/include/CService.h]: Fixed.

File: Server/BackEndMonitoringServer/include/Utils.h (hunk 6 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/include/Utils.h]: This files useless, deleted now.

File: Server/BackEndMonitoringServer/src/CCommandLineHandler.cpp (hunk 7 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/src/CCommandLineHandler.cpp]: Thnx, changed to switch-case

File: Server/BackEndMonitoringServer/src/CService.cpp (hunk 8 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/src/CService.cpp:23]: Added \\Todo

File: Server/BackEndMonitoringServer/src/CService.cpp (hunk 9 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/src/CService.cpp]: Formatted

File: Server/BackEndMonitoringServer/src/CServiceHandler.cpp (hunk 10 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/src/CServiceHandler.cpp]: Edited

File: Server/BackEndMonitoringServer/src/CServiceHandler.cpp (hunk 11 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/src/CServiceHandler.cpp:61]: Ok, @Serwlashev please, edit this in Sockets branch.

File: Server/BackEndMonitoringServer/src/CServiceHandler.cpp (hunk 12 above)
Comment (VytalyGorbatov) [reply Server/BackEndMonitoringServer/src/CServiceHandler.cpp]: Fixed

File: Sockets/BackEndMonitoringSockets/include/CAcceptorWrapper.h (hunk 13 above)
Comment (VytalyGorbatov) [reply Sockets/BackEndMonitoringSockets/include/CAcceptorWrapper.h:27]: @Serwlashev, this too.

File: Sockets/BackEndMonitoringSockets/src/CAcceptor.cpp (hunk 14 above)
Comment (VytalyGorbatov) [reply Sockets/BackEndMonitoringSockets/src/CAcceptor.cpp:24]: @Serwlashev, also this)

File: Sockets/BackEndMonitoringSockets/src/CAcceptorWrapper.cpp (hunk 15 above)
Comment (VytalyGorbatov) [reply Sockets/BackEndMonitoringSockets/src/CAcceptorWrapper.cpp:11]: @Serwlashev, and this

File: Server/BackEndMonitoringServer/include/CService.h [hunk 16]
```diff
@@ -5,11 +5,11 @@
 
 struct ServiceParameters
 {
-    const CString& name         = "BackendMonitoringService";
+    const CString& name = "BackendMonitoringService";
```
Comment (AntonVoytov) [inline Server/BackEndMonitoringServer/include/CService.h:8]: Replace all CString with std::string.
const CString& name = "BackendMonitoringService"; ->  const std::string name("BackendMonitoringService"); if possible or init via ctor.
Check other places.

File: Server/BackEndMonitoringServer/src/CService.cpp [hunk 17]
```diff
@@ -62,31 +63,31 @@ DWORD WINAPI CService::ServiceCtrlHandler(
...
+        m_p_service->Stop();
     }
 
     return 0;
 }
 
 void WINAPI CService::SvcMain(DWORD argc, CHAR** argv) 
 {
-    assert(m_service);
+    assert(m_p_service);
```
Comment (AntonVoytov) [inline Server/BackEndMonitoringServer/src/CService.cpp:74]: Do not assert, use conditional instead.

File: Server/BackEndMonitoringServer/src/CService.cpp [hunk 18]
```diff
@@ -62,31 +63,31 @@ DWORD WINAPI CService::ServiceCtrlHandler(
...
-    m_service->Start(argc, argv);
+    m_p_service->Start(argc, argv);
 }
 
 bool CService::Run()
 {
-    m_service = this;
+    m_p_service = this;
 
     CHAR* name = const_cast<CString&>(m_name).GetBuffer();
```
Comment (AntonVoytov) [inline Server/BackEndMonitoringServer/src/CService.cpp:92]: CHAR* -> char*

**Review by AntonVoytov — COMMENTED** [review]
Please fix and reply to each comment.

File: Server/BackEndMonitoringServer/src/CProcessInfo.cpp [hunk 19]
```diff
@@ -0,0 +1,173 @@
...
+}
+
+EMemoryConvertType CProcessInfo::GetMemoryCountType() const
+{ return m_count_type;}
+
+
+bool CProcessInfo::CountCpuUsage()
+{
+    unsigned long long cur_sys_time, cur_kernel_time, cur_user_time;
+    cur_sys_time = cur_kernel_time = cur_user_time = 0;
```
Comment (AntonVoytov) [inline Server/BackEndMonitoringServer/src/CProcessInfo.cpp:140]: Do not do that! It looks disgusting. Initialize simple types when declare.

File: utils/include/WindowsUtils.cpp [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+#include "PlatformUtils.h"
+
+#pragma warning(disable : 6385)
+
+namespace PlatformUtils
+{
+	bool GetExistingProcessIds(std::vector<unsigned>& container_of_PIDs)
+	{
+		unsigned short m_max_process_count = 1024;
+		DWORD* p_process_ids = new DWORD[m_max_process_count];
```
Comment (AntonVoytov) [inline utils/include/WindowsUtils.cpp:12]: Try to use a smart pointer for a array

File: utils/include/WindowsUtils.cpp [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+	}
+
+	bool GetProcessTimes(unsigned PID, unsigned long long& system_time,
+						 unsigned long long& kernel_time, 
+						 unsigned long long& user_time)
+	{
+		HANDLE process = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ,
+			FALSE, PID);
+
+		bool success = (process != 0);
```
Comment (AntonVoytov) [inline utils/include/WindowsUtils.cpp:47]: Check what HANDLE type is actually. It should be void* or void**

File: utils/include/Utils.h [hunk 22]
```diff
@@ -1,16 +1,28 @@
...
+
 namespace Utils
 {
 	/// <summary>
 	/// Return current date and time in system
 	/// </summary>
 	/// <returns>format: dd.mm.yyyy hh:mm:ss</returns>
-	bool GetCurrentDateAndTimeFormatted(std::string& 
-		date_time_var_to_save);
+	[[nodiscard]] bool TryGetCurrentDateAndTimeFormatted(std::string&
```
Comment (AntonVoytov) [inline utils/include/Utils.h:11]: If you prefer to use a single return, try to use do {} while construction.

**Review by AntonVoytov — COMMENTED** [review]
Please check comments

**Pull request closed without merging**